src/dialogs/PreviewUploadOverlay.cc

```diff
@@ -0,0 +1,181 @@
+        fileName_.setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
+        fileName_.setAlignment(Qt::AlignCenter);
+        upload_.setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+        cancel_.setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+        upload_.setFontSize(conf::btn::fontSize);
+        cancel_.setFontSize(conf::btn::fontSize);
+
+        if (isImage_) {
+                infoLabel_.setAlignment(Qt::AlignCenter);
+
+                // Scale image preview to the size of the current window if it is larger.
+                if ((img_size.height() * img_size.width()) > (winsize.height() * winsize.width())) {
+                        infoLabel_.setPixmap(image_.scaled(winsize, Qt::KeepAspectRatio));
+                } else {
+                        infoLabel_.setPixmap(image_);
+                        move(center.x() - (width() * 0.5), center.y() - (height() * 0.5));
+                }
+        } else {
+                infoLabel_.setAlignment(Qt::AlignLeft);
+        }
+        infoLabel_.setScaledContents(false);
+
+        show();
+}
+
+void
+PreviewUploadOverlay::setPreview(const QByteArray data, const QString &mime)
+{
+        auto const &split = mime.split('/');
+        auto const &type  = split[1];
+
+        data_      = data;
+        mediaType_ = split[0];
+        filePath_  = "clipboard." + type;
+        isImage_   = false;
+
+        if (mediaType_ == "image") {
+                if (!image_.loadFromData(data_)) {
+                        titleLabel_.setText(QString{tr(ERROR)}.arg(type));
+                } else {
+                        titleLabel_.setText(QString{tr(DEFAULT)}.arg(mediaType_));
+                }
+                isImage_ = true;
+        } else {
+                auto const info = QString{"Media type: %1\n"
+                                          "Media size: %2\n"}
+                                    .arg(mime)
+                                    .arg(utils::humanReadableFileSize(data.size()));
+
+                titleLabel_.setText(QString{tr(DEFAULT)}.arg("file"));
+                infoLabel_.setText(tr(info.toLocal8Bit()));
+        }
+
+        init();
+}
+
+void
+PreviewUploadOverlay::setPreview(const QString &path)
+{
+        QFile file{path};
+
+        if (!file.open(QIODevice::ReadOnly)) {
+                qWarning() << "Failed to open file from:" << path;
+                qWarning() << "Reason:" << file.errorString();
+                close();
+                return;
+        }
+
+        QMimeDatabase db;
+        auto mime = db.mimeTypeForFileNameAndData(path, &file);
+
+        if ((data_ = file.readAll()).isEmpty()) {
+                qWarning() << "Failed to read media:" << file.errorString();
+                close();
+                return;
+        }
+
+        auto const &split = mime.name().split('/');
+
+        mediaType_ = split[0];
+        filePath_  = file.fileName();
+        isImage_   = false;
+
+        if (mediaType_ == "image") {
```

Comment:
Everything below here should be factored out into a common function called by both overloads.

Reply:
Will do!